### PR TITLE
Adds DM/GM limit based on pref

### PR DESCRIPTION
**Pull request by shazm**
#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note

```

### CODE REVIEW

**Comment by enahum** [comment]
@shazm do we need to filter DM & GMs in other categories too? What about the last time there were msgs in the channel? I ask cause I still have a GM in my favs that is not visible on the webapp

**Comment by shazm** [comment]
Screen Shot 2022-03-21 at 10 27 12 am

Good question - that's the setting itself. I'm not sure if it implies that's the total number of direct messages to show across all categories, or just the number of channels to show in the Direct Messages category. 

As the description states; you can set it here as well:

Screen Shot 2022-03-21 at 10 30 49 am

Which to me implies that you only set that number for that category.

From what you describe, the behaviour is the former on the webapp - I might refer to @esethna / @matthewbirtch to make that call on what it should be.

**Comment by enahum** [comment]
> 
> From what you describe, the behaviour is the former on the webapp - I might refer to @matthewbirtch to make that call on what it should be.

I don't think is the former as it currently is not displayed on the webapp on daily but it is with this PR.

I would encourage you to look at the current webapp and try and match every functionality

**Comment by shazm** [comment]
> @shazm do we need to filter DM & GMs in other categories too? What about the last time there were msgs in the channel? I ask cause I still have a GM in my favs that is not visible on the webapp

> I don't think is the former as it currently is not displayed on the webapp on daily but it is with this PR.

Ah! I assumed you were trying to say that GMs/DMs should be limited across all categories (based on the user pref) - the behaviour you describe is probably a separate issue to this PR then; as the webapp is following the latter behaviour which I've implemented here (just limit the number of channels in the direct messages category).

File: app/components/channel_list/categories/body/index.ts [hunk 1]
```diff
@@ -87,7 +88,21 @@ const enhance = withObservables(['category'], ({category, locale, database}: {ca
...
 
+    let limit = of$('');
+    if (category.type === 'direct_messages') {
+        limit = database.get<PreferenceModel>(PREFERENCE).
+            query(
+                Q.where('category', Preferences.CATEGORY_SIDEBAR_SETTINGS),
+                Q.where('name', 'limit_visible_dms_gms'),
+            ).observe().pipe(
+                switchMap(
+                    (val) => of$(val[0].value),
```
Comment (enahum) [inline app/components/channel_list/categories/body/index.ts:99]: This should be checking if val[0] exist if not return of$('')

Also I rather do the parseInt here and have the component receive an integer.

**Review by enahum — CHANGES_REQUESTED** [review]
Based on the convo this is the only change, do we have tickets to look at the entire behavior regardless of the category to match the webapp?

**Comment by shazm** [comment]
> Based on the convo this is the only change, do we have tickets to look at the entire behavior regardless of the category to match the webapp?

Added: https://mattermost.atlassian.net/browse/MM-42641

**Review by enahum — APPROVED** [review]

File: app/components/channel_list/categories/body/category_body.tsx [hunk 2]
```diff
@@ -27,9 +30,13 @@ const CategoryBody = ({currentChannelId, sortedIds, category}: Props) => {
         );
     }, [currentChannelId]);
 
+    if (category.type === 'direct_messages' && limit > 0) {
+        ids = sortedIds.slice(0, limit - 1);
```
Comment (enahum) [inline app/components/channel_list/categories/body/category_body.tsx:34]: Actually you should useMemo for `ids` as is always a new array

**Review by enahum — CHANGES_REQUESTED** [review]

**Review by larkox — APPROVED** [review]
Apart of enahum concern, LGTM.

**Comment by matthewbirtch** [comment]
> I'm not sure if it implies that's the total number of direct messages to show across all categories, or just the number of channels to show in the Direct Messages category.

Just chiming in here to confirm that it's the number to show in the Direct Messages category. Doesn't apply to the other categories

**Review by enahum — APPROVED** [review]

**Pull request merged**